### PR TITLE
refactored handler for adding new routes

### PR DESCRIPTION
**Pull request by Joseph-Cashmore**
Alex held my hand...not really because Covid.

### CODE REVIEW

File: .github/workflows/build_and_deploy.yml [hunk 1]
```diff
@@ -29,14 +29,13 @@ jobs:
...
-          go get -v -t -d ./src/...
+        run: go get -v -t -d
 
       - name: Build
-        run: GOARCH=amd64 GOOS=linux go build -o main ./src
+        run: GOARCH=amd64 GOOS=linux go build -o main
 
       - name: Test
-        run: go test -v ./src/
+        run: go test -v ./...
```
Comment (samstradling) [inline .github/workflows/build_and_deploy.yml:38]: Have we checked if this can be:
```suggestion
        run: make test
```

Comment (alexNaylorTeece) [reply .github/workflows/build_and_deploy.yml]: Yes, done

File: template.yaml [hunk 2]
```diff
@@ -14,8 +14,8 @@ Resources:
   HelloWorldFunction:
     Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
     Properties:
-      CodeUri: src/
-      Handler: src
+      CodeUri: ./
+      Handler: .
```
Comment (samstradling) [inline template.yaml:18]: To get this to work I needed to make the following changes:
```diff
diff --git a/template.yaml b/template.yaml
index 6fa5908..5731a10 100644
--- a/template.yaml
+++ b/template.yaml
@@ -15,15 +15,15 @@ Resources:
     Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
     Properties:
       CodeUri: ./
-      Handler: .
+      Handler: main
       Runtime: go1.x
       Tracing: Active # https://docs.aws.amazon.com/lambda/latest/dg/lambda-x-ray.html
       Events:
         CatchAll:
           Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
           Properties:
-            Path: /hello
-            Method: GET
+            Path: /{proxy+}
+            Method: ANY
       Environment: # More info about Env Vars: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#environment-object
         Variables:
           PARAM1: VALUE
```

File: main.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	if result != nil {
+		var err error
+		res, err = json.Marshal(result)
+		if err != nil {
+			return events.APIGatewayProxyResponse{
+				Body:       err.Error(),
+				StatusCode: 500,
+			}, nil
+		}
+	}
```
Comment (samstradling) [inline main.go:29]: This part seems a bit confusing to me, if `result == nil` we end up with an empty body (which isn't valid json), but we're still sending a response content type of `application/json`:

```bash
$ curl -vs http://127.0.0.1:3000/hi 2>&1 | grep Content-
< Content-Type: application/json
< Content-Length: 0
```

However, if we remove the `result != nil` check:

```diff
diff --git a/main.go b/main.go
index 5dccf86..fb7cf93 100644
--- a/main.go
+++ b/main.go
@@ -15,17 +15,13 @@ type handler struct {
 
 func (h *handler) doRequest(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
 	result, statusCode := h.router.Route(request)
+	res, err := json.Marshal(result)
 
-	var res []byte
-	if result != nil {
-		var err error
-		res, err = json.Marshal(result)
-		if err != nil {
-			return events.APIGatewayProxyResponse{
-				Body:       err.Error(),
-				StatusCode: 500,
-			}, nil
-		}
+	if err != nil {
+		return events.APIGatewayProxyResponse{
+			Body:       err.Error(),
+			StatusCode: 500,
+		}, nil
 	}
 
 	return events.APIGatewayProxyResponse{

```

Then we get `null` as the body which is at least valid json. Alternatively we could change the response content-type.

**Review by samstradling — APPROVED** [review]

File: src/main.go [hunk 4]
```diff
@@ -1,30 +0,0 @@
-package main
```
[the remark this answers is not in the document]
Comment (alexNaylorTeece) [reply src/main.go:1]: I guess no real reason other than the fact that I have never seen a repo where the main wasn't at top level before. Happy to move it back - no strong feeling either way

File: main.go (hunk 3 above)
Comment (alexNaylorTeece) [reply main.go:29]: Yeah good point - will change

**Pull request merged**